pkg/virt-handler/device-manager/generic_device.go

```diff
@@ -21,223 +21,75 @@ package device_manager
-	devs         []*pluginapi.Device
-	server       *grpc.Server
-	socketPath   string
-	stop         <-chan struct{}
-	health       chan deviceHealth
-	devicePath   string
-	deviceName   string
-	resourceName string
-	done         chan struct{}
-	deviceRoot   string
-	preOpen      bool
-	initialized  bool
-	lock         *sync.Mutex
-	permissions  string
-	deregistered chan struct{}
+	*DevicePluginBase
+	preOpen     bool
+	permissions string
 }
 
 func NewGenericDevicePlugin(deviceName string, devicePath string, maxDevices int, permissions string, preOpen bool) *GenericDevicePlugin {
 	serverSock := SocketPath(deviceName)
 	dpi := &GenericDevicePlugin{
-		devs:         []*pluginapi.Device{},
-		socketPath:   serverSock,
-		health:       make(chan deviceHealth),
-		deviceName:   deviceName,
-		devicePath:   devicePath,
-		deviceRoot:   util.HostRootMount,
-		resourceName: fmt.Sprintf("%s/%s", DeviceNamespace, deviceName),
-		preOpen:      preOpen,
-		initialized:  false,
-		lock:         &sync.Mutex{},
-		permissions:  permissions,
+		DevicePluginBase: &DevicePluginBase{
+			devs:         []*pluginapi.Device{},
+			socketPath:   serverSock,
+			deviceRoot:   util.HostRootMount,
+			devicePath:   devicePath,
+			health:       make(chan deviceHealth),
+			done:         make(chan struct{}),
+			deregistered: make(chan struct{}),
+			resourceName: fmt.Sprintf("%s/%s", DeviceNamespace, deviceName),
+			initialized:  false,
+			lock:         &sync.Mutex{},
+		},
+		preOpen:     preOpen,
+		permissions: permissions,
 	}
 
 	for i := 0; i < maxDevices; i++ {
-		deviceId := dpi.deviceName + strconv.Itoa(i)
+		deviceId := deviceName + strconv.Itoa(i)
 		dpi.devs = append(dpi.devs, &pluginapi.Device{
 			ID:     deviceId,
-			Health: pluginapi.Healthy,
+			Health: pluginapi.Unhealthy,
 		})
 	}
-
+	dpi.GetIDDeviceName = dpi.GetIDDeviceNameFunc
+	dpi.SetupMonitoredDevices = dpi.SetupMonitoredDevicesFunc
+	dpi.SetupDevicePlugin = dpi.SetupDevicePluginFunc
 	return dpi
 }
 
-func (dpi *GenericDevicePlugin) GetDeviceName() string {
-	return dpi.deviceName
-}
-
-// Start starts the device plugin
-func (dpi *GenericDevicePlugin) Start(stop <-chan struct{}) (err error) {
-	logger := log.DefaultLogger()
-	dpi.stop = stop
-	dpi.done = make(chan struct{})
-	dpi.deregistered = make(chan struct{})
-
-	err = dpi.cleanup()
-	if err != nil {
-		return err
-	}
-
+func (dpi *GenericDevicePlugin) SetupDevicePluginFunc() error {
 	// The kernel module(s) for some devices, like tun and vhost-net, auto-load when needed.
 	// That need is identified by the first access to their main device node.
 	// Opening and closing the device nodes here will trigger any necessary modprobe.
 	if dpi.preOpen {
-		devnode, err := os.Open(dpi.devicePath)
+		devicePath := filepath.Join(dpi.deviceRoot, dpi.devicePath)
+		devnode, err := os.Open(devicePath)
 		if err == nil {
 			devnode.Close()
+		} else {
+			return fmt.Errorf("error opening the device node %s: %v", devicePath, err)
```

Comment:
Also stylistic nit: Could be an early return and then just call `devnode.Close()`

```suggestion
		if err != nil {
			return fmt.Errorf("error opening the device node %s: %v", devicePath, err)
		}
		devnode.Close()
```